### ISSUE TITLE
USB plug-and-play support

### ISSUE THREAD

**Issue by rsta2** [rsta2/circle#129]
An initial attempt has been made to add USB plug-and-play (PnP) support to Circle. It is available on the branch *usb-pnp*.

There is a test program attached to this comment, which demonstrates USB PnP with a standard USB keyboard. It is the modified sample/08-usbkeyboard. The keyboard can be connected or removed at any time while the system is running.

Other devices are currently not fully supported for PnP. Attachments and removes to/from the USB root port (e.g. on Raspberry Pi Zero, A, A+, 3A+) are currently not handled. All other RPis have an USB hub connected to the root port and should work.

The system option *USE_USB_SOF_INTR* should be enabled in the file *include/circle/sysconfig.h* to use USB PnP on RPi 1-3.

[usb-pnp-test.zip](https://github.com/rsta2/circle/files/5149469/usb-pnp-test.zip)

**Comment by rsta2** [comment]
Attachments and removes to/from the USB root port (e.g. on Raspberry Pi Zero, A, A+, 3A+) should be detected now too.

**Comment by rsta2** [comment]
The following sample programs and related USB devices are plug-and-play enabled now on the *usb-pnp* branch:

* 08-usbkeyboard
* 10-usbmouse
* 27-usbgamepad
* 29-miniorgan
* 37-showgamepad

**Comment by rsta2** [comment]
USB plug-and-play support has been completed for now and has been merged into the *develop* branch. There is a new documentation file [doc/usb-plug-and-play.txt](https://github.com/rsta2/circle/blob/develop/doc/usb-plug-and-play.txt), which describes details of the USB PnP support. Unfortunately I detected a cache-alignment problem of DMA buffers, while testing the USB PnP support, which has to be solved before the next Circle version can be released.
